rust/src/composepost.rs

```diff
@@ -28,6 +29,123 @@ use crate::treefile::Treefile;
 const RPMOSTREE_RPMDB_LOCATION: &str = "usr/share/rpm";
 const TRADITIONAL_RPMDB_LOCATION: &str = "var/lib/rpm";
 
+fn dir_renameat(
+    src: &openat::Dir,
+    srcname: &str,
+    destdir: &openat::Dir,
+    destname: &str,
+) -> Result<()> {
+    nix::fcntl::renameat(
+        Some(src.as_raw_fd()),
+        srcname,
+        Some(destdir.as_raw_fd()),
+        destname,
+    )?;
+    Ok(())
+}
+
+#[context("Moving {}", name)]
+fn dir_move_if_exists(src: &openat::Dir, dest: &openat::Dir, name: &str) -> Result<()> {
+    if src.exists(name)? {
+        dir_renameat(src, name, dest, name)?;
+    }
+    Ok(())
+}
+
+/// Initialize an ostree-oriented root filesystem.
+///
+/// This is hardcoded; in the future we may make more things configurable,
+/// but the goal is for all state to be in `/etc` and `/var`.
+#[context("Init rootfs")]
+pub(crate) fn compose_init_rootfs(
+    rootfs_dfd: &openat::Dir,
+    treefile: &mut Treefile,
+) -> CxxResult<()> {
+    use nix::fcntl::OFlag;
+    println!("Initializing rootfs");
+
+    // Unfortunately fchmod() doesn't operate on an O_PATH descriptor
+    let flag = OFlag::O_DIRECTORY | OFlag::O_CLOEXEC;
+    let nix_rootfs = nix::dir::Dir::openat(
+        rootfs_dfd.as_raw_fd(),
+        ".",
+        flag,
+        Mode::from_bits(0o755).unwrap(),
```

Comment:
This shows up in a few place, so it can probably be moved to a `const` using `from_bits_truncate` or bit ORing.

Reply:
Using the API to make it properly via `|` is *very* verbose, and truncating feels like it can hide problems.  Just went with a shared variable.